### DISCUSSION TITLE
New Query feature does not behave well with pydantic AliasGenerator

### DISCUSSION THREAD

**Discussion by rich-young-slalom** [fastapi/fastapi#12224] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated, Literal

from fastapi import FastAPI, Query
from pydantic import BaseModel, Field, ConfigDict, AliasGenerator
from pydantic.alias_generators import to_camel

app = FastAPI()


class FilterParams(BaseModel):
    model_config = ConfigDict(
        alias_generator=AliasGenerator(validation_alias=to_camel), extra="forbid"
    )
    
    limit: int = Field(100, gt=0, le=100)
    offset: int = Field(0, ge=0)
    order_by: Literal["created_at", "updated_at"] = "created_at"
    other_tags: list[str] = []


@app.get("/items/")
async def read_items(filter_query: Annotated[FilterParams, Query()]):
    return filter_query
```

### Description

My request models have an alias generator using the default to_camel alias generator for validation. This allows me to conform to api requirements of camel case while my python models use snake case.

For the above request for example, I would want to be able to call the endpoint like

```shell
/items/?orderBy=updated_at&otherTags=1&otherTags=2
```

However I am encountering a number of issues. To begin with, the defaults aren't playing nice.

If I call 

```shell
/items/
``` 

I will get an extra_forbidden on field order_by (and if I remove the default I will get an extra_forbidden on field other_tags)

If I add populate_by_name=True to the ConfigDict I can then run the endpoint successfully which seems to imply that fastapi is pushing a dict to the pydantic using the default names. 

```python
class FilterParams(BaseModel):
    model_config = ConfigDict(
        alias_generator=AliasGenerator(validation_alias=to_camel), extra="forbid", populate_by_name=True
    )
    
    limit: int = Field(100, gt=0, le=100)
    offset: int = Field(0, ge=0)
    order_by: Literal["created_at", "updated_at"] = "created_at"
    other_tags: list[str] = []
``` 

Worth noting that if the default value is None - I don't have this issue, so again I assume fastapi is stripping null values from the dict.

i.e.

```python 
order_by: Literal["created_at", "updated_at"] | None = None
``` 

=================================================================

For query requests there are two issues.

If I have added populate_by_name to the ConfigDict and then call

```shell
\items\?orderBy=updated_at
```

I will now get an extra_forbidden error on order_by with input "created_at" again as it looks like it's passing in the defined query parameter as orderBy correcly BUT ALSO the default value thanks to populate_by_name

Again, if I have no defaults or the defaults are None this works without issue.

The second issue is that lists do not work. If I call

```shell
\items\?otherTags=1
```

I will get an error that input should be a valid list with input being 1

if I call 

```shell
\items\?otherTags=1&otherTags=2
```

I will get an error that input should be a valid list with input being 2 (so appears to only include the last one)

=================================================================

It is also important to note that both these issues only affect snake cased fields. 

If you renamed the fields to orderby or othertags, but kept the configdict as is, everything would work as expected. 

For this reason I can get around this by naming the fields as their camelCase names, however I have run POST and other requests using the AliasGenerator. and it has worked perfectly, not to mention it also works perfectly for non list fields that have a default value of None (or are required). 

### Operating System

Linux

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

3.11.6

**Comment by tonnico** [comment]
I have the same issues described here: https://github.com/fastapi/fastapi/discussions/12231

If it is intended, I would close my question and would move my minimal examples in this one?

**Comment by YuriiMotov** [comment]
Several issues related to aliases was fixed in FastAPI 0.124.4.
I just checked, this works well in current version!
